File: bcw4/Basic/Complex/Complex.cpp
#include <iostream>

#include "Complex.h"

Complex::Complex(double real, double imaginary) : x(0), y(0) {
    this->real = real;
    this->imaginary = imaginary;
}

Complex::~Complex() {}

double Complex::getReal() const {
    return real;
}

double Complex::getImaginary() const {
    return imaginary;
}

bool Complex::operator==(const Complex& other) const {
    return (real == other.real && imaginary == other.imaginary);
}

bool Complex::operator!=(const Complex& other) const {
    return (real != other.real || imaginary != other.imaginary);
}

void Complex::operator+=(const Complex& other) {
    real += other.real;
    imaginary += other.imaginary;
}
void Complex::operator-=(const Complex& other) {
    real -= other.real;
    imaginary -= other.imaginary;
}

Complex Complex::operator+(const Complex& other) const {
    Complex temp;
    
    temp.real = real + other.real;
    temp.imaginary = imaginary + other.imaginary;
    return temp;
}

Complex Complex::operator-(const Complex& other) const {
    Complex temp;
    
    temp.real = real - other.real;
    temp.imaginary = imaginary - other.imaginary;
    return temp;
}

Complex Complex::operator*(const Complex& other) const {
    Complex temp;
    
    temp.real = real * other.real - imaginary * other.imaginary;
    temp.imaginary = imaginary * other.real + real * other.imaginary;
    return temp;
}

std::ostream& operator<< (std::ostream& out, const Complex& complex) {
    if ( complex.getImaginary() >= 0 ) {
        out << complex.getReal() << "+" << complex.getImaginary() << "i";
        return out;
    }
    out << complex.getReal() << complex.getImaginary() << "i";
    return out;
}

File: bcw4/java/Basic/Complex/Complex.java
public class Complex {
    private double real;
    private double imaginary;

    public Complex() {
        this.real = 0.0;
        this.imaginary = 0.0;
    }

    public Complex(double real, double imaginary) {
        this.real = real; 
        this.imaginary = imaginary;
    }

    public double getReal() {
        return real;
    }
    
    public double getImaginary() {
        return imaginary;
    }

    public boolean equals(Complex other) {
        return real == other.real && imaginary == other.imaginary;
    }

    public boolean notEquals(Complex other) {
        return !equals(other);
    }

    public void addTo(Complex other) {
        real += other.real;
        imaginary += other.imaginary;
    }

    public void subtractFrom(Complex other) {
        real -= other.real;
        imaginary -= other.imaginary;
    }

    public Complex sum(Complex other) {
        Complex temp = new Complex();

        temp.real = real + other.real;
        temp.imaginary = imaginary + other.imaginary;
        return temp;
    }

    public Complex subtraction(Complex other) {
        Complex temp = new Complex();

        temp.real = real - other.real;
        temp.imaginary = imaginary - other.imaginary;
        return temp;   
    }

    public Complex multiplication(Complex other) {
        Complex temp = new Complex();
    
        temp.real = real * other.real - imaginary * other.imaginary;
        temp.imaginary = imaginary * other.real + real * other.imaginary;
        return temp;
    }

    public String toString() {
        if ( imaginary >= 0 ) {
            return  real + "+" + imaginary + "i";
        }
        return  real + "" + imaginary + "i";
    }

    public void print() {
        if ( imaginary >= 0) {
            System.out.format("%.2f+%.2fi%n", real, imaginary);  
        } else {
            System.out.format("%.2f%.2fi%n", real, imaginary);
        }
    }
}